x-pack/plugins/observability_solution/apm_data_access/server/utils/unflatten_known_fields.ts

```diff
@@ -0,0 +1,167 @@
+  SPAN_LINKS_SPAN_ID,
+  SPAN_LINKS_TRACE_ID,
+  ...CONCRETE_FIELDS
+} = APM_EVENT_FIELDS_MAP;
+
+const ALL_FIELDS = Object.values(CONCRETE_FIELDS);
+
+const KNOWN_MULTI_VALUED_FIELDS = [
+  APM_EVENT_FIELDS_MAP.CHILD_ID,
+  APM_EVENT_FIELDS_MAP.PROCESS_ARGS,
+] as const;
+
+type KnownField = ValuesType<typeof CONCRETE_FIELDS>;
+
+type KnownSingleValuedField = Exclude<KnownField, KnownMultiValuedField>;
+type KnownMultiValuedField = ValuesType<typeof KNOWN_MULTI_VALUED_FIELDS>;
+
+const KNOWN_SINGLE_VALUED_FIELDS = ALL_FIELDS.filter(
+  (field): field is KnownSingleValuedField => !KNOWN_MULTI_VALUED_FIELDS.includes(field as any)
+);
+
+interface TypeOverrideMap {
+  [APM_EVENT_FIELDS_MAP.SPAN_DURATION]: number;
+  [APM_EVENT_FIELDS_MAP.AGENT_NAME]: AgentName;
+  [APM_EVENT_FIELDS_MAP.EVENT_OUTCOME]: EventOutcome;
+  [APM_EVENT_FIELDS_MAP.FAAS_COLDSTART]: true;
+  [APM_EVENT_FIELDS_MAP.TRANSACTION_DURATION]: number;
+  [APM_EVENT_FIELDS_MAP.TIMESTAMP_US]: number;
+  [APM_EVENT_FIELDS_MAP.PROCESSOR_EVENT]: ProcessorEvent;
+  [APM_EVENT_FIELDS_MAP.SPAN_COMPOSITE_COUNT]: number;
+  [APM_EVENT_FIELDS_MAP.SPAN_COMPOSITE_SUM]: number;
+  [APM_EVENT_FIELDS_MAP.SPAN_SYNC]: boolean;
+  [APM_EVENT_FIELDS_MAP.TRANSACTION_SAMPLED]: boolean;
+  [APM_EVENT_FIELDS_MAP.PROCESSOR_NAME]: 'transaction' | 'metric' | 'error';
+  [APM_EVENT_FIELDS_MAP.HTTP_RESPONSE_STATUS_CODE]: number;
+  [APM_EVENT_FIELDS_MAP.PROCESS_PID]: number;
+  [APM_EVENT_FIELDS_MAP.OBSERVER_VERSION_MAJOR]: number;
+  [APM_EVENT_FIELDS_MAP.ERROR_EXC_HANDLED]: boolean;
+}
+
+type MaybeMultiValue<T extends KnownField, U> = T extends KnownMultiValuedField ? U[] : U;
+
+type TypeOfKnownField<T extends KnownField> = MaybeMultiValue<
+  T,
+  T extends keyof TypeOverrideMap ? TypeOverrideMap[T] : string
+>;
+
+type MapToSingleOrMultiValue<T extends Record<string, any>> = {
+  [TKey in keyof T]: TKey extends KnownField
+    ? T[TKey] extends undefined
+      ? TypeOfKnownField<TKey> | undefined
+      : TypeOfKnownField<TKey>
+    : unknown;
+};
+
+type UnflattenedKnownFields<T extends Record<string, any>> = DedotObject<
+  MapToSingleOrMultiValue<T>
+>;
+
+export type FlattenedApmEvent = Record<KnownSingleValuedField | KnownMultiValuedField, unknown[]>;
+
+export type UnflattenedApmEvent = UnflattenedKnownFields<FlattenedApmEvent>;
+
+export function unflattenKnownApmEventFields<T extends Record<string, any> | undefined = undefined>(
+  fields: T
+): T extends Record<string, any> ? UnflattenedKnownFields<T> : undefined;
+
+export function unflattenKnownApmEventFields<
+  T extends Record<string, any> | undefined,
+  U extends Array<keyof Exclude<T, undefined>>
+>(
+  fields: T,
+  required: U
+): T extends Record<string, any>
+  ? UnflattenedKnownFields<T> &
+      (U extends any[]
+        ? UnflattenedKnownFields<{
+            [TKey in ValuesType<U>]: keyof T extends TKey ? T[TKey] : unknown[];
+          }>
+        : {})
+  : undefined;
+
+export function unflattenKnownApmEventFields(
```

Comment:
ideally we have tests here as well - there's also a test_types file IIRC that validates some assumptions about the types, I think it would be useful to add that as well.

Reply:
Do you mean increasing the coverage of [this](https://github.com/elastic/kibana/pull/195242/files/5756999ffe4f545f72df71daee37adc25b286907#diff-0ab94c0f4ca8a5fbc0c513f28bb94f96d64a7c50926a22a2971679452b5fad65)

Reply:
no I just misssed it, but let me comment there